rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,658 @@
+                .await
+                .map_err(|_| StorageError::S3Error)?;
+
+            if let Some(identifiers) = self.unpack_object_identifier(&list_obj_resp) {
+                self.delete_objects(bucket, identifiers)
+                    .await
+                    .map_err(|_| StorageError::S3Error)?;
+            } else {
+                break Ok(());
+            }
+
+            // check if more objects exists
+            continuation_token = self.unpack_next_continuation_token(&list_obj_resp);
+            if continuation_token.is_none() {
+                break Ok(());
+            }
+        }
+    }
+
+    // Get the object identifier/keys
+    fn unpack_object_identifier(
+        &self,
+        list_obj_resp: &ListObjectsV2Output,
+    ) -> Option<Vec<ObjectIdentifier>> {
+        if let Some(objects) = &list_obj_resp.contents {
+            let keys = objects
+                .into_iter()
+                .filter_map(|obj| obj.key.clone())
+                .map(|key| ObjectIdentifier {
+                    key,
+                    ..Default::default()
+                })
+                .collect();
+            Some(keys)
+        } else {
+            None
+        }
+    }
+
+    // Delete the objects of the given bucket
+    async fn delete_objects(
+        &self,
+        bucket: &str,
+        identifiers: Vec<ObjectIdentifier>,
+    ) -> Result<DeleteObjectsOutput, RusotoError<DeleteObjectsError>> {
+        let req = DeleteObjectsRequest {
+            bucket: bucket.to_string(),
+            delete: Delete {
+                objects: identifiers,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        self.s3_client.delete_objects(req).await
+    }
+
+    /// Return all objects keys for the given bucket.
+    async fn list_objects(
+        // the response returns 1000 keys max.
+        // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property
+        &self,
+        bucket: &str,
+        continuation_token: Option<String>,
+    ) -> Result<ListObjectsV2Output, RusotoError<ListObjectsV2Error>> {
+        // list all objects
+        let req = ListObjectsV2Request {
+            bucket: bucket.to_string(),
+            continuation_token,
+            // Minio is not limited to 1000
+            max_keys: Some(1000),
+            ..Default::default()
+        };
+
+        self.s3_client.list_objects_v2(req).await
+    }
+
+    // Unpack the next_continuation_token of the ListObjectsV2Output response
+    fn unpack_next_continuation_token(
+        &self,
+        list_obj_resp: &ListObjectsV2Output,
+    ) -> Option<String> {
+        // https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingJava.html
+        if let Some(is_truncated) = list_obj_resp.is_truncated {
+            if is_truncated {
+                list_obj_resp.next_continuation_token.clone()
```

Comment:
Is it possible that `is_truncated` is `true` but `next_continuation_token` is `false`? That won't affect the returned value here, but if that occurs, we might silently abort iterating through a truncated object in the caller.

Reply:
That should not be possible but we should test this😁.
In the [rusoto](https://docs.rs/rusoto_s3/0.43.0/rusoto_s3/struct.ListObjectsV2Output.html#structfield.next_continuation_token) documentation they wrote: 

> NextContinuationToken is sent when isTruncated is true, which means there are more keys in the bucket that can be listed. The next list requests to Amazon S3 can be continued with this NextContinuationToken. NextContinuationToken is obfuscated and is not a real key

Reply:
Perhaps is worth checking that this holds true and just log an error if it doesn't then.